Update FrequencyMapAgent and tests to match documentation

dianarg opened the pull request:
- Uncore can not be set to fixed frequency through the policy.
- NAN is not allowed for mapped regions.
- Removed consideration of hints.
- Fixed up integration test.
- Resolves #1052.
- Fixes #1009.

Change-Id: Ia09fe225d939cce5c3ae6c9ccf42d6e2b96dac2c
Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>
Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Update FrequencyMapAgent and tests to match documentation

- Uncore can not be set to fixed frequency through the policy.
- NAN is not allowed for mapped regions.
- Removed consideration of hints.
- Fixed up integration test.
- Resolves #1052.
- Fixes #1009.

Change-Id: Ia09fe225d939cce5c3ae6c9ccf42d6e2b96dac2c
Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>
Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

bgeltz commented on ronn/geopm_agent_frequency_map.7.ronn:
```diff
@@ -73,7 +73,9 @@ name (see **geopm(7)**).  This name can also be passed to the
-                  NAN will result in a runtime error.
+                  NAN will result in a runtime error, unless the policy
+                  is all NAN, in which case the agent will not exert
+                  any control.
```
Why allow an all NAN policy?  Why have this agent not exert control?  We have the monitor for this use case.

cmcantalupo replied:
This issue is around the case where a non-root controller starts up and has not yet received a policy from it's parent.  In this case, we don't want the agent to do anything until a policy with values arrives.

bgeltz replied:
If a NAN policy comes through the Endpoint, this is not an error.  We need to consider this use case.

Do we expect the policy writer to actually write an all NAN's policy?

If the agent gets a non-all-NAN's policy, from that point forward should the agent throw if it receives an all-NAN policy?

Change the agent to throw that it will not accept an all NANs policy after it has received a valid policy.  Eventually even the initial all-NAN policy should be an error (this should be a TODO in the source).  Remove the documentation that all-NANs is valid from the man page.

bgeltz commented on ronn/geopm_agent_frequency_map.7.ronn:
```diff
@@ -124,10 +126,8 @@ raised.
-policy default CPU frequency).
-
+The assigned frequency is added to each region's section of the report
+if that region was explicitly assigned a frequency in the policy map.
```
Why remove the feature of having frequencies for all regions show up in the report, including those not in the policy map?

cmcantalupo replied:
The problem is that the Agent interface is kind of not awesome.  The agent is not given a list of all regions to report on, it has to observe them in a sample to know about them.  For this reason, we will not be able to provide the keys for the map returned by the Agent::report_region() api:  https://github.com/geopm/geopm/blob/dev/src/Agent.hpp#L121 for short running regions. In at least the case of a short running region not in the policy map, the agent is unable to meet the requirement.  So if it isn't reliably there, I said why put there for the unmapped regions.

bgeltz replied:
Please make it so that if any region runs long enough to get the default frequency from the policy, that the default frequency shows up in the individual region's section in the report (not just up in the top policy section).

dianarg committed:
Fix tests and address review comments

- geopm_test_launcher must always be imported if used outside of
  --skip-launch.
- All-NAN policy is only valid if received before a real policy.  In
  the future, the controller should be fixed so that this does not
  happen.

Change-Id: I35ff9a04c81a75c50c1ec3990be045492b12bd0f
Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>

bgeltz commented on src/FrequencyMapAgent.cpp:
```diff
@@ -169,36 +167,52 @@ namespace geopm
         for (auto it = policy.begin() + M_POLICY_FIRST_HASH;
-             it != policy.end() && std::next(it) != policy.end(); std::advance(it, 2)) {
+             it != policy.end() && std::next(it) != policy.end();
+             std::advance(it, 2)) {
```
Why?

Also, it looks like there's a dangling curly brace at EOL.

dianarg replied:
This is just an extra newline.  The advance by two is because this part of the policy has pairs of hash + frequency.

cmcantalupo replied:
Why advance two elements?  One is the hash one is the frequency.  Not sure about the dangling curly, it was there before the change.

bgeltz commented on src/FrequencyMapAgent.cpp:
```diff
@@ -169,36 +167,52 @@ namespace geopm
+                // validate policy, which the controller should
+                // call before this function.
+                GEOPM_DEBUG_ASSERT(!std::isnan(freq),
+                                   "mapped region with no frequency assigned.");
```
Shouldn't it be impossible to hit this assert?

dianarg replied:
Yes if you are using the controller; since this is a public interface, you can still hit it through unit testing.

cmcantalupo replied:
That is why it is a debug assert.

bgeltz commented on src/FrequencyMapAgent.cpp:
```diff
@@ -169,36 +167,52 @@ namespace geopm
+            m_is_policy_updated = true;
+            m_default_freq = policy[M_POLICY_FREQ_DEFAULT];
+        }
+        if (m_hash_freq_map != old_freq_map) {
```
I'm surprised this works.

dianarg replied:
Me too!

cmcantalupo replied:
?

bgeltz commented on src/FrequencyMapAgent.cpp:
```diff
@@ -242,71 +256,83 @@ namespace geopm
+        if (!m_is_adjust_initialized) {
+            // adjust all controls once in case not applied by policy
+            for (size_t ctl_idx = 0; ctl_idx < (size_t) m_num_freq_ctl_domain; ++ctl_idx) {
+                // @todo: this is bad; agent should be able to use aliases
```
Is there a bug open?

dianarg replied:
Yes: #795

bgeltz commented on test_integration/test_frequency_hint_usage.py:
```diff
@@ -50,10 +50,11 @@
 import geopmpy.error
 
 from test_integration import util
+from test_integration import geopm_test_launcher
```
Why?

If this is wrong, it's wrong in the template too.

cmcantalupo replied:
The problem is that the test fixtures use the launcher in this test to get the min and max frequencies.  This is not generally going to be the case.

bgeltz commented on test_integration/test_frequency_hint_usage.py:
```diff
@@ -98,13 +99,13 @@ def setUpClass(cls):
         cls._fmap_report_path = 'test_{}_fmap.report'.format(test_name)
         cls._fmap_trace_path = 'test_{}_fmap.trace'.format(test_name)
-        cls._fmap_agent_conf_path = test_name + '-fmap-agent-config.json'
+        cls._fmap_agent_conf_path = 'test_' + test_name + '-fmap-agent-config.json'
```
Wrong in the template too.

bgeltz replied:
Bug opened for this: #1090

bgeltz commented on test_integration/test_frequency_hint_usage.py:
```diff
@@ -98,13 +99,13 @@ def setUpClass(cls):
+        cls._ee_agent_conf_path = 'test_' + test_name + '-ee-agent-config.json'
 
-        geopmpy.error.exc_clear() # Clear out exception record for python 2 support
+        geopmpy.error.exc_clear()  # Clear out exception record for python 2 support
```
Why the space?

dianarg replied:
pretty sure this is a pep8 style guideline; my linter was complaining.

bgeltz commented on test_integration/test_frequency_hint_usage.py:
```diff
@@ -138,7 +139,7 @@ def setUpClass(cls):
             launcher.set_num_node(num_node)
             launcher.set_num_rank(num_rank)
-            launcher.run(test_name)
+            launcher.run('test_' + test_name)
```
Wrong in the template too.

bgeltz commented on test_integration/test_frequency_hint_usage.py:
```diff
@@ -207,12 +206,13 @@ def test_frequency_sane_ee(self):
+        self.assertNotEqual(self._freq_min, assigned_freq,
+                            msg='Expected assigned frequency to not be the min from policy since this region has the COMPUTE hint. ({} == {}).'.format(self._freq_min, assigned_freq))
+        util.assertNotNear(self, self._freq_min, achieved_freq,
+                           msg='Expected achieved frequency to NOT be near min from policy. ({} ~= {})'.format(self._freq_min, achieved_freq))
```
I don't see anything changed here besides putting msg on it's own line.

bgeltz commented on test_integration/test_frequency_hint_usage.py:
```diff
@@ -207,12 +206,13 @@ def test_frequency_sane_ee(self):
 if __name__ == '__main__':
     # Call do_launch to clear non-pyunit command line option
     util.do_launch()
     unittest.main()
```
This newline is vital. :P

I'm noticing a trend where patches I write introduce stuff like this, and patches that Chris does remove them...
